### ISSUE TITLE
Client side navigation to SSR page executes middleware only once

### ISSUE THREAD

**Issue by SergeyZubatkin** [vercel/next.js#30938]
### What version of Next.js are you using?

12.0.2

### What version of Node.js are you using?

14.x

### What browser are you using?

Chrome

### What operating system are you using?

Windows

### How are you deploying your application?

Vercel

### Describe the Bug

I have app deployed to Vercel with all pages using SSR.
middleware is used for randomly rewrite **/ssr/home** to the **a** or **b** page content.
The issue is that middleware executes only once, during prefetching, subsequent client navigation for the same page does not execute middleware.

Scenario on deployed app:
Open Network Tab
Navigate to https://ab-testing-simple-8up4jxiuu-sergeyzubatkin.vercel.app/ssr
On Network you can see that middleware for **ssr/home** is executed
Click on the **ssr/home** link and then **Home Page SSR** link. 
On Network there is calls for new json (correct as it is SSR) BUT there is no HEAD calls to execute middleware.

### Expected Behavior

Middleware is called for each click on the SSR page link.

### To Reproduce

Public repo with source code https://github.com/SergeyZubatkin/ab-testing-issue

**Comment by javivelasco** [comment]
This is expected since we are caching `HEAD` requests to reduce the amount of requests as much as possible which can still be problematic (#30901). Although I'm aware it is possible that the request input can change since we first ran the HEAD request and therefore the result can change, I'm curious if this is causing any particular issue in your case or is it just that you expected to be executed again.

This "preflight" checks work in a slightly different way from SSR. While those happen synchronously on navigation we tried not to run those the same way. Maybe we can provide configuration for this cc @timneutkens @ijjk

**Comment by SergeyZubatkin** [comment]
@javivelasco Thanks for information.
Our simplified use case we want to support:
Site with 3 pages: /home /marketer /developer
We want to build personalization on the /home page with some targeted content.
After user visits /developer page we want to use developer specific content on home page. 
Personalization should be performed on the edge and be compatible with client side navigation.

**Comment by javivelasco** [comment]
Closing this in favour of https://github.com/vercel/next.js/issues/32727

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
